activerecord/lib/active_record/associations/builder/belongs_to.rb

```diff
@@ -31,9 +31,10 @@ def self.add_counter_cache_methods(mixin)
         def belongs_to_counter_cache_after_update(reflection)
           foreign_key  = reflection.foreign_key
           cache_column = reflection.counter_cache_column
+          @_after_create_counter_called ||= {}
 
-          if (@_after_create_counter_called ||= false)
-            @_after_create_counter_called = false
+          if (@_after_create_counter_called[foreign_key] ||= false)
```

Comment:
I agree with you.  I was unwisely making decisions for an easier to read diff instead of easier to read code.  I've pushed up a new version.